src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionTests.cs

```diff
@@ -124,20 +125,34 @@ public async Task PumpAllAsync_WhenUnsupportedEvent_Ignores()
     }
 
     [TestMethod]
-    public async Task PumpAllAsync_WhenPublisherDisposed_Finishes()
+    public async Task PumpAllAsync_WhenPublisherErrors_NonCriticalError_ErrorLoggedAndSessionIsDisposed()
     {
         var testScope = new TestScope();
         testScope.SetUpSwitchToBackgroundThread();
         var sseStreamMock = testScope.SetUpSQServiceToSuccessfullyReturnSSEStreamReader();
-        testScope.SetUpSSEStreamReaderToReturnEventsSequenceAndExit(sseStreamMock, 
-            new IServerEvent[]{Mock.Of<ITaintVulnerabilityRaisedServerEvent>(), Mock.Of<ITaintVulnerabilityRaisedServerEvent>()});
+        sseStreamMock.Setup(x => x.ReadAsync()).Throws(new NotImplementedException("this is a test"));
 
-        testScope.TaintPublisherMock.Setup(publisher => publisher.Publish(It.IsAny<ITaintServerEvent>()))
-            .Throws(new ObjectDisposedException(string.Empty));
+        testScope.LoggerMock.Setup(x => x.LogVerbose(It.Is<string>(s => s.Contains("this is a test")), Array.Empty<object>()));
 
         await testScope.TestSubject.PumpAllAsync();
 
-        testScope.TaintPublisherMock.Verify(publisher => publisher.Publish(It.IsAny<ITaintServerEvent>()), Times.Once);
+        testScope.LoggerMock.Verify(x=> x.LogVerbose(It.Is<string>(s=> s.Contains("this is a test")), Array.Empty<object>()), Times.Once);
+        testScope.CapturedSessionToken.Value.IsCancellationRequested.Should().BeTrue();
+    }
+
+    [TestMethod]
+    public async Task PumpAllAsync_WhenPublisherErrors_CriticalError_SessionThrows()
+    {
+        var testScope = new TestScope();
+        testScope.SetUpSwitchToBackgroundThread();
+        var sseStreamMock = testScope.SetUpSQServiceToSuccessfullyReturnSSEStreamReader();
+        sseStreamMock.Setup(x => x.ReadAsync()).Throws(new StackOverflowException("this is a test"));
```

Comment:
That's probably one of the worst exceptions to actually use as an example:

```
Starting with 2.0 a StackOverflow Exception can only be caught in the following circumstances.

The CLR is being run in a hosted environment* where the host specifically allows for StackOverflow exceptions to be handled
The stackoverflow exception is thrown by user code and not due to an actual stack overflow situation ([Reference](https://learn.microsoft.com/en-us/archive/blogs/jaredpar/when-can-you-catch-a-stackoverflowexception))
*"hosted environment" as in "my code hosts CLR and I configure CLR's options" and not "my code runs on shared hosting"
```

Choose a different one please